cmd/zoekt-sourcegraph-indexserver/index.go

```diff
@@ -160,30 +165,145 @@ func gitIndex(o *indexArgs, runCmd func(*exec.Cmd) error) error {
 		return err
 	}
 
-	fetchStart := time.Now()
+	var fetchCommits = func(commits []string) error {
+		fetchArgs := []string{"-C", gitDir, "-c", "protocol.version=2", "fetch", "--depth=1", o.CloneURL}
+		fetchArgs = append(fetchArgs, commits...)
+
+		cmd = exec.CommandContext(ctx, "git", fetchArgs...)
+		cmd.Stdin = &bytes.Buffer{}
+
+		return runCmd(cmd)
+	}
 
 	// We shallow fetch each commit specified in zoekt.Branches. This requires
 	// the server to have configured both uploadpack.allowAnySHA1InWant and
 	// uploadpack.allowFilter. (See gitservice.go in the Sourcegraph repository)
-	fetchArgs := []string{"-C", gitDir, "-c", "protocol.version=2", "fetch", "--depth=1", o.CloneURL}
-	var commits []string
+	var branchCommits []string
 	for _, b := range o.Branches {
-		commits = append(commits, b.Version)
+		branchCommits = append(branchCommits, b.Version)
 	}
-	fetchArgs = append(fetchArgs, commits...)
 
-	cmd = exec.CommandContext(ctx, "git", fetchArgs...)
-	cmd.Stdin = &bytes.Buffer{}
+	commitCount := 0
+	commitCount += len(branchCommits)
 
-	err = runCmd(cmd)
+	fetchStart := time.Now()
+	err = fetchCommits(branchCommits)
 	fetchDuration := time.Since(fetchStart)
+
 	if err != nil {
 		metricFetchDuration.WithLabelValues("false", repoNameForMetric(o.Name)).Observe(fetchDuration.Seconds())
 		return err
 	}
 
+	if o.UseDelta {
```

Comment:
This feels like logic that should all live inside zoekt-git-index.

Is this all done due to fetchCommits? If so I'd prefer a much smaller function footprint here that looks something like

```
if o.UseDelta {
  repo, ok, err := buildOptions.Current()
  if err != nil {
    // hard fail
  }
  if ok {
    err := fetchCommits(repo.RepositoryBranches)
    if err != nil {
       // log error don't fail since this means the branch was force pushed/etc.
    }
  }
}

// Current returns the currently indexed api.Repository on disk. If not indexed ok is false.
```

Any smarts around falling back to not using delta should be done inside of zoekt-git-index. In indexserver it should only be concerned with setting things up for delta. I imagine delta as the only mode in the future from an outsiders perspective, so imagine indexserver not having any `o.UseDelta` branches as our final state. The zoekt-git-index essentially uses delta if it can.

Reply:
Yes, this is all for fetch commits. The idea was to be defensive and verify the delta invariant and fail early if we can (since we have to loop over all the branches in the existing shard anyway in order to `git fetch` the associated commits). 


---

buildOptions.Current seems similar to https://github.com/sourcegraph/zoekt/pull/259. Perhaps it's time to pull that in?

Reply:
pulling in sounds good